Sources/Currency/AnyCurrency+Algorithms.swift

```diff
@@ -0,0 +1,87 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Currency open source project
+//
+// Copyright (c) 2020 Currency project authors
+// Licensed under MIT License
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Currency project authors
+//
+// SPDX-License-Identifier: MIT
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+// MARK: Value Distribution
+
+extension AnyCurrency {
+  /// Distributes the current amount into a set number of parts as evenly as possible.
+  /// - Note: Passing a negative or `0` value will result in an empty result.
+  /// - Complexity: O(1)
+  /// - Parameter numParts: The count of new values the single value should be distributed between as evenly as possible.
+  /// - Returns: A collection of currency values with their share of the amount distribution.
+  public func distributedEvenly(intoParts numParts: Int) -> [Self] {
+    guard numParts > 0 else { return [] }
+
+    let count = Int64(numParts)
+    
+    // courtesy of https://codereview.stackexchange.com/a/221221
+    let units = self.minorUnits
+    let fraction = units / count
+    let remainder = Int(abs(units) % count)
+    
+    var results: [Self] = .init(repeating: Self(exactly: 0), count: numParts)
+    for index in 0..<remainder {
+      results[index] = Self(exactly: fraction + units.signum())
+    }
+    for index in remainder..<numParts {
+      results[index] = Self(exactly: fraction)
+    }
+    
+    return results
+  }
+  
+  /// Distributes the current amount between other amounts proportionally based on their original value.
+  ///
+  /// The resulting amounts will match the same sign (negative or positive) as the amount being distributed.
+  ///
+  /// For example:
+  ///
+  ///     let result = USD(-10).distributedProportionally(between: [5, 8.25])
+  ///     // result == [USD(-3.77), USD(-6.23)]
+  ///
+  /// - Complexity: O(2*n*), where *n* is the number of `originalValues`.
+  /// - Parameter originalValues: A collection of values that should be scaled proportionally so that their sum equals this currency's amount.
+  /// - Returns: A collection of currency values that are scaled proportionally from an original value whose sum equals this currency's amount.
+  public func distributedProportionally<C: Collection>(
+    between originalValues: C
+  ) -> [Self] where C.Element == Self {
+    guard originalValues.count > 0 else { return [] }
```

Comment:
With example: `USD(30.30).distributedProportionally(between: [])`

1) We return no elements, because you didn't give anything to distribute against
2) We return `[USD(30.30)]`

I could see an argument for both, and don't have a solid opinion on either.

Reply:
I see it both ways too, let's keep this for now.